API/hermes/hermes.cpp

```diff
@@ -1446,11 +1449,84 @@ jsi::Value HermesRuntimeImpl::evaluateJavaScript(
-        "bootstrap"
-    );
+
+    auto preBootstrap = std::string("nativeModuleProxy.AliucordNative.getConstants()");
+    auto constants = evaluateJavaScript(std::make_unique<jsi::StringBuffer>(preBootstrap), "pre-bootstrap")
+                             .asObject(*this);
+    auto bootstrapPath = constants.getProperty(*this, "externalStorageDirectory")
+                            .asString(*this).utf8(*this) + "/AliucordRN/bootstrap.js";
+
+    std::string bootstrap;
+
+    if (access(bootstrapPath.c_str(), F_OK) == 0) {
+      ::hermes::hermesLog(
+          "AliuHermes",
+          "Loading bootstrap from external %s",
+          bootstrapPath.c_str());
+
+      auto file = fopen(bootstrapPath.c_str(), "rb");
+      if (!file) {
+        ::hermes::hermesLog(
+            "AliuHermes",
+            "Failed to open external bootstrap: %s",
+            strerror(errno));
+        return jsi::Value::null();
+      }
+
+      fseek(file, 0, SEEK_END);
+      int size = ftell(file);
+      fseek(file, 0, SEEK_SET);
+
+      char *data = (char*)malloc(size);
+      if (!fgets(data, size, file)) {
+        ::hermes::hermesLog(
+            "AliuHermes",
+            "Failed to read specified number of bytes: %s",
+            strerror(errno));
+        fclose(file);
+        return jsi::Value::null();
+      }
+      fclose(file);
+
+      bootstrap = std::string(data);
+    } else {
+      ::hermes::hermesLog("AliuHermes", "Loading bootstrap from APK");
+
+      auto packageCodePath = constants.getProperty(*this, "packageCodePath")
+                                 .asString(*this).utf8(*this);
+
+      auto zip = zip_open(packageCodePath.c_str(), 0, 'r');
+      if (LLVM_UNLIKELY(!zip)) {
+        ::hermes::hermesLog(
+            "AliuHermes",
+            "Failed to open packageCodePath to read bootstrap");
+        return jsi::Value::null();
+      }
+
+      if (zip_entry_open(zip, "bootstrap.js") != 0) {
+        ::hermesLog(
+            "AliuHermes",
+            "Failed to open bootstrap.js in packageCodePath: %s ",
+            strerror(errno));
+        return jsi::Value::null();
+      }
+      size_t size = zip_entry_size(zip);
+
+      void *data = nullptr;
+      if (LLVM_UNLIKELY(zip_entry_read(zip, &data, &size) < 0)) {
+        ::hermes::hermesLog(
+            "AliuHermes",
+            "Failed to read bootstrap.js from packageCodePath: %s",
+            strerror(errno));
+        return jsi::Value::null();
+      }
+      zip_entry_close(zip);
+      zip_close(zip);
+
+      bootstrap = std::string(::hermes::vm::ASCIIRef((char *)data, size).data());
```

Comment:
same as above, you very likely need to free the zip entry, either via free or maybe there's a special zip_free